datafusion/functions-aggregate/src/min_max.rs

```diff
@@ -15,103 +15,165 @@
+    LargeBinaryArray, LargeStringArray, StringArray, Time32MillisecondArray,
+    Time32SecondArray, Time64MicrosecondArray, Time64NanosecondArray,
+    TimestampMicrosecondArray, TimestampMillisecondArray, TimestampNanosecondArray,
+    TimestampSecondArray, UInt16Array, UInt32Array, UInt64Array, UInt8Array,
+};
 use arrow::compute;
 use arrow::datatypes::{
-    DataType, Date32Type, Date64Type, IntervalUnit, Time32MillisecondType,
-    Time32SecondType, Time64MicrosecondType, Time64NanosecondType, TimeUnit,
-    TimestampMicrosecondType, TimestampMillisecondType, TimestampNanosecondType,
-    TimestampSecondType,
-};
-use arrow::{
-    array::{
-        ArrayRef, BinaryArray, BooleanArray, Date32Array, Date64Array, Float32Array,
-        Float64Array, Int16Array, Int32Array, Int64Array, Int8Array,
-        IntervalDayTimeArray, IntervalMonthDayNanoArray, IntervalYearMonthArray,
-        LargeBinaryArray, LargeStringArray, StringArray, Time32MillisecondArray,
-        Time32SecondArray, Time64MicrosecondArray, Time64NanosecondArray,
-        TimestampMicrosecondArray, TimestampMillisecondArray, TimestampNanosecondArray,
-        TimestampSecondArray, UInt16Array, UInt32Array, UInt64Array, UInt8Array,
-    },
-    datatypes::Field,
+    DataType, Decimal128Type, Decimal256Type, Float32Type, Float64Type, Int16Type,
+    Int32Type, Int64Type, Int8Type, UInt16Type, UInt32Type, UInt64Type, UInt8Type,
 };
-use arrow_array::types::{
-    Decimal128Type, Float32Type, Float64Type, Int16Type, Int32Type, Int64Type, Int8Type,
-    UInt16Type, UInt32Type, UInt64Type, UInt8Type,
-};
-use arrow_array::{BinaryViewArray, StringViewArray};
-use datafusion_common::internal_err;
-use datafusion_common::ScalarValue;
-use datafusion_common::{downcast_value, DataFusionError, Result};
-use datafusion_expr::{Accumulator, GroupsAccumulator};
-
-use crate::aggregate::utils::down_cast_any_ref;
-use crate::expressions::format_state_name;
-use arrow::array::Array;
-use arrow::array::Decimal128Array;
-use arrow::array::Decimal256Array;
+use arrow_schema::IntervalUnit;
+use datafusion_common::{downcast_value, internal_err, DataFusionError, Result};
+use datafusion_physical_expr_common::aggregate::groups_accumulator::prim_op::PrimitiveGroupsAccumulator;
+use std::fmt::Debug;
+
 use arrow::datatypes::i256;
-use arrow::datatypes::Decimal256Type;
+use arrow::datatypes::{
+    Date32Type, Date64Type, Time32MillisecondType, Time32SecondType,
+    Time64MicrosecondType, Time64NanosecondType, TimeUnit, TimestampMicrosecondType,
+    TimestampMillisecondType, TimestampNanosecondType, TimestampSecondType,
+};
 
-use super::moving_min_max;
+use datafusion_common::ScalarValue;
+use datafusion_expr::{
+    function::AccumulatorArgs, Accumulator, AggregateUDFImpl, Signature, Volatility,
+};
+use datafusion_expr::{type_coercion, Expr, GroupsAccumulator};
+
+pub static STRINGS: &[DataType] = &[DataType::Utf8, DataType::LargeUtf8];
```

Comment:
These data types are not used, we could remove them

Reply:
Thanks, this is a remainder of a first effort to get it working. I had question about the signature that is required for min max and eventually settling on `Signature:user_defined` work, but this isn't what originally was in the 
https://github.com/apache/datafusion/blob/main/datafusion/expr/src/aggregate_function.rs#L117-L130

I migrated this signature to the new version, and nothing worked, then I realized that the adjacent reimplementation of `sum` udf uses a different signature...so I adopted it, and all tests passd.

This is the only change that I find mysterious and I am not really convinced about

Reply:
Actually udaf and builtin goes to different path. Builtin in `coerce_agg_exprs_for_signature`, UDAF in `coerce_arguments_for_signature_with_aggregate_udf`. Directly moving the builtin signature to UDAF make things complicated.

After you coerce with `user_defined`, `coerce_types` tell you what is actually happening. It makes coercion of the function a lot easier and trivial. Although ideally, this is only for function that has specialized need, others function should use datafusion's builtin signature (which sadly is not easy to use yet).

The `coerce_types` is called [here](https://github.com/apache/datafusion/blob/3fe18604e8a127dbecf78acff0dce7596790bfe8/datafusion/expr/src/type_coercion/functions.rs#L190-L194)

You can trace the code to understand how the data types are coerced and send to `return_type`.